usr.bin/paste/paste.c

```diff
@@ -98,51 +114,50 @@ main(int argc, char *argv[])
 		delim = tab;
 	}
 
-	if (seq)
-		rval = sequential(argv);
-	else
-		rval = parallel(argv);
+	for (filecnt = 0; *argv; ++argv, ++filecnt) {
+		failed = 0;
+		if ((lp = malloc(sizeof(LIST))) == NULL)
+			err(EXIT_FAILURE, NULL);
+		if ((*argv)[0] == '-' && !(*argv)[1])
+			lp->fp = stdin;
+		else if (!(lp->fp = fopen(*argv, "r"))) {
+			if (!seq)
+				err(EXIT_FAILURE, "%s", *argv);
+			else
+				failed = errno;
+		}
+
+		if (cap_rights_limit(fileno(lp->fp),
+		    cap_rights_init(&rights, CAP_READ)) < 0 &&
+		    errno != ENOSYS)
+			err(EXIT_FAILURE, "unable to limit rights for %s", *argv);
+
+		lp->cnt = filecnt;
+		lp->err = failed;
+		lp->name = *argv;
+
+		STAILQ_INSERT_TAIL(&lh, lp, entries);
+	}
+
+	if (cap_enter() < 0 && errno != ENOSYS)
+		err(EXIT_FAILURE, "failed to enter capability mode");
+
+	rval = seq ? sequential() : parallel();
+
 	exit(rval);
 }
 
-typedef struct _list {
-	struct _list *next;
-	FILE *fp;
-	int cnt;
-	char *name;
-} LIST;
-
 static int
-parallel(char **argv)
+parallel(void)
 {
 	LIST *lp;
-	int cnt;
 	wint_t ich;
 	wchar_t ch;
-	char *p;
-	LIST *head, *tmp;
 	int opencnt, output;
 
-	for (cnt = 0, head = tmp = NULL; (p = *argv); ++argv, ++cnt) {
-		if ((lp = malloc(sizeof(LIST))) == NULL)
-			err(1, NULL);
-		if (p[0] == '-' && !p[1])
-			lp->fp = stdin;
-		else if (!(lp->fp = fopen(p, "r")))
-			err(1, "%s", p);
-		lp->next = NULL;
-		lp->cnt = cnt;
-		lp->name = p;
-		if (!head)
-			head = tmp = lp;
-		else {
-			tmp->next = lp;
-			tmp = lp;
-		}
-	}
-
-	for (opencnt = cnt; opencnt;) {
-		for (output = 0, lp = head; lp; lp = lp->next) {
+	for (opencnt = filecnt; opencnt;) {
```

Comment:
In opencnt can we do a logical expression?

Reply:
Do you mean using `opencnt > 0`? That’s certainly a valid approach and it’s a good idea if you want to make the code more self-explanatory. It avoids relying on implicit truthiness, which can sometimes be less clear. Personally, I prefer to keep the code brief and concise where possible, which is why I didn’t use that form. However, if you think it would be helpful for clarity, I’d be open to considering the change.

Reply:
Yes, I prefer to have a boolean expression in conditions it makes it easier to understand.

Reply:
Roger that, I'll take care of it.